tests/model/ChangeSetTest.php

```diff
@@ -0,0 +1,439 @@
+<?php
+
+/**
+ * Provides a set of targettable permissions for tested models
+ *
+ * @mixin Versioned
+ * @mixin DataObject
+ */
+trait ChangeSetTest_Permissions {
+	public function canEdit($member = null) {
+		return $this->can(__FUNCTION__, $member);
+	}
+
+	public function canDelete($member = null) {
+		return $this->can(__FUNCTION__, $member);
+	}
+
+	public function canCreate($member = null, $context = array()) {
+		return $this->can(__FUNCTION__, $member, $context);
+	}
+
+	public function canPublish($member = null, $context = array()) {
+		return $this->can(__FUNCTION__, $member, $context);
+	}
+
+	public function canUnpublish($member = null, $context = array()) {
+		return $this->can(__FUNCTION__, $member, $context);
+	}
+
+	public function can($perm, $member = null, $context = array()) {
+		$perms = [
+			"PERM_{$perm}",
+			'CAN_ALL',
+		];
+		return Permission::checkMember($member, $perms);
+	}
+}
+
+/**
+ * @mixin Versioned
+ */
+class ChangeSetTest_Base extends DataObject implements TestOnly {
+	use ChangeSetTest_Permissions;
+
+	private static $db = [
+		'Foo' => 'Int',
+	];
+
+	private static $has_many = [
+		'Mids' => 'ChangeSetTest_Mid',
+	];
+
+	private static $owns = [
+		'Mids',
+	];
+
+	private static $extensions = [
+		"Versioned",
+	];
+}
+
+/**
+ * @mixin Versioned
+ */
+class ChangeSetTest_Mid extends DataObject implements TestOnly {
+	use ChangeSetTest_Permissions;
+
+	private static $db = [
+		'Bar' => 'Int',
+	];
+
+	private static $has_one = [
+		'Base' => 'ChangeSetTest_Base',
+		'End' => 'ChangeSetTest_End',
+	];
+
+	private static $owns = [
+		'End',
+	];
+
+	private static $extensions = [
+		"Versioned",
+	];
+}
+
+/**
+ * @mixin Versioned
+ */
+class ChangeSetTest_End extends DataObject implements TestOnly {
+	use ChangeSetTest_Permissions;
+
+	private static $db = [
+		'Baz' => 'Int',
+	];
+
+	private static $extensions = [
+		"Versioned",
+	];
+}
+
+/**
+ * Test {@see ChangeSet} and {@see ChangeSetItem} models
+ */
+class ChangeSetTest extends SapphireTest {
+
+	protected static $fixture_file = 'ChangeSetTest.yml';
+
+	protected $extraDataObjects = [
+		'ChangeSetTest_Base',
+		'ChangeSetTest_Mid',
+		'ChangeSetTest_End',
+	];
+
+	/**
+	 * Automatically publish all objects
+	 */
+	protected function publishAllFixtures() {
+		$this->logInWithPermission('ADMIN');
+		foreach($this->fixtureFactory->getFixtures() as $class => $fixtures) {
+			foreach ($fixtures as $handle => $id) {
+				$this->objFromFixture($class, $handle)->doPublish();
```

Comment:
`doPublish()` is explicitly for publishing outside of changesets. it's the "automatic publish" feature.

`publish()` is the low level api that is used by both changesets and doPublish() that actually does all the work. The difference is that publish() can be used to publish in either direction between draft and live, and that it doesn't attempt to do recursive publishing.

ChangeSets provide the users an ability to group unrelated changes together in an atomic way, which doPublish() cannot do. It also acts as a "trusted" action, as all changes in a changeset are considered vetted and approved by a human operator, whereas doPublish() is automatic and has to make more decisions itself, thus is less "trusted".